canary/producer-cpp/src/Canary.cpp

```diff
@@ -0,0 +1,899 @@
+}
+
+STATUS
+CanaryStreamCallbackProvider::droppedFrameReportHandler(UINT64 custom_data, STREAM_HANDLE stream_handle,
+                                                        UINT64 dropped_frame_timecode) {
+    LOG_WARN("Reporting dropped frame. Frame timecode " << dropped_frame_timecode);
+    return STATUS_SUCCESS;
+}
+
+STATUS
+CanaryStreamCallbackProvider::fragmentAckReceivedHandler(UINT64 custom_data, STREAM_HANDLE stream_handle,
+                                                         UPLOAD_HANDLE upload_handle, PFragmentAck pFragmentAck) {
+    CustomData *data = reinterpret_cast<CustomData *>(custom_data);
+    if (pFragmentAck->ackType != FRAGMENT_ACK_TYPE_PERSISTED && pFragmentAck->ackType != FRAGMENT_ACK_TYPE_RECEIVED)
+    {
+        return STATUS_SUCCESS;
+    }
+
+    cout << "Map Debug: searching in map for a key of pFragmentAck->timestamp: " << pFragmentAck->timestamp << endl;
+    map<uint64_t, uint64_t>::iterator iter;
+    iter = data->timeOfNextKeyFrame->find(pFragmentAck->timestamp);
+    if(iter == data->timeOfNextKeyFrame->end())
+    {
+        cout << "Map Debug: TimeOfNextKeyFrame key-value pair not present in map" << endl;
+    }
+    else
+    {
+        cout << "Map Debug: Found TimeOfNextKeyFrame key-value pair in map" << endl;
+        uint64_t timeOfFragmentEndSent = data->timeOfNextKeyFrame->find(pFragmentAck->timestamp)->second;
+        cout << "Map Debug: with value of " << timeOfFragmentEndSent << endl;
+        cout << "Map Debug: latency: " << duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count() - timeOfFragmentEndSent << endl;
+
+        // When Canary sleeps, timeOfFragmentEndSent become less than currentTimeStamp, don't send those metrics
+        if (timeOfFragmentEndSent > pFragmentAck->timestamp)
+        {
+            if (pFragmentAck->ackType == FRAGMENT_ACK_TYPE_PERSISTED)
+            {
+                Aws::CloudWatch::Model::MetricDatum persistedAckLatency_datum;
+                Aws::CloudWatch::Model::PutMetricDataRequest cwRequest;
+                cwRequest.SetNamespace("KinesisVideoSDKCanaryCPP");
+
+                auto currentTimestamp = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
+                auto persistedAckLatency = (currentTimestamp - timeOfFragmentEndSent); // [milliseconds]
+                pushMetric("PersistedAckLatency", persistedAckLatency, Aws::CloudWatch::Model::StandardUnit::Milliseconds, persistedAckLatency_datum, data->pDimension_per_stream, cwRequest);
+                LOG_DEBUG("Persisted Ack Latency: " << persistedAckLatency);
+                if (data->pCanaryConfig->useAggMetrics)
+                {
+                    pushMetric("PersistedAckLatency", persistedAckLatency, Aws::CloudWatch::Model::StandardUnit::Milliseconds, persistedAckLatency_datum, data->pAggregated_dimension, cwRequest);
+
+                }
+                auto outcome = data->pCWclient->PutMetricData(cwRequest);
```

Comment:
Let us change to an async call. Use PutMetricDataAsync instead. We do not want to block on this. You can refer to producer C canary for reference.